### PR TITLE
Fix & test git_identifer

### PR DESCRIPTION
**Pull request by encukou**
Alas, what's not tested is broken.
The function misbehaved with capital letters, and a few edge cases.
Let's throw some property-based testing at it.

### CODE REVIEW

File: naucse/compiled_renderer.py [hunk 1]
```diff
@@ -194,14 +194,14 @@ def replacement(match):
             return result
         num = ord(char)
         if num < 2**8:
-            return '-c{num:02x}'
```
Comment (hroncok) [inline naucse/compiled_renderer.py:197]: Sorry for not spotting this during the initial review :blush:

File: tox.ini [hunk 2]
```diff
@@ -2,5 +2,6 @@
 envlist = py38,py39,310
 
 [testenv]
+extras = dev
```
Comment (hroncok) [inline tox.ini:5]: This circulary requires tox in toxenv. I don't think the way the test depepndencies are defined is sane. pytest is defined in the extra nad also in deps.

What about:

 - \[test\] extra with pytest and hypothesis
 - \[dev\] extra with tox and the above as well (or not at all)
 - adapt the readme about executing pytest directly to pip install `.[test]` first

**Review by hroncok — COMMENTED** [review]
The tests look fine and are a lovely example of how to use hypothesis. Thanks. Left some minor comments.

**Comment by encukou** [comment]
Thinking about it on the way to lunch, I decided to make the output more readable (for URLs), similarly to how punycode works.
Need to stop yak-shaving now.

File: tox.ini [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [tox]
-envlist = py38,py39,310
+envlist = py38,py39,310,311
```
Comment (hroncok) [inline tox.ini:2]: FTR the added Python 3.11 does not run on the CI.

**Comment by hroncok** [comment]
Si there any way to see the tested strings in pytest output? It seems only 6 tests are listed in the output.

**Comment by encukou** [comment]
> Si there any way to see the tested strings in pytest output? It seems only 6 tests are listed in the output.

Not really – Hypothesis tries a bunch of them at random.
You could add a `print` and run with `-s`

**Review by hroncok — APPROVED** [review]

**Comment by encukou** [comment]
Thank you!
